crates/biome_html_formatter/src/utils/children.rs

```diff
@@ -0,0 +1,422 @@
+use std::{
+    iter::{FusedIterator, Peekable},
+    str::Chars,
+};
+
+use biome_formatter::{
+    format_args, prelude::*, write, Buffer, Format, FormatElement, FormatResult,
+};
+use biome_html_syntax::AnyHtmlElement;
+use biome_rowan::{SyntaxResult, TextLen, TextRange, TextSize, TokenText};
+
+use crate::{context::HtmlFormatContext, HtmlFormatter};
+
+pub(crate) static HTML_WHITESPACE_CHARS: [char; 4] = [' ', '\n', '\t', '\r'];
+
+/// Meaningful HTML text is defined to be text that has either non-whitespace
+/// characters, or does not contain a newline. Whitespace is defined as ASCII
+/// whitespace.
+///
+/// ```
+/// use biome_html_formatter::utils::children::is_meaningful_html_text;
+///
+/// assert_eq!(is_meaningful_html_text("     \t\r   "), true);
+/// assert_eq!(is_meaningful_html_text("     \n\r   "), false);
+/// assert_eq!(is_meaningful_html_text("  Alien   "), true);
+/// assert_eq!(is_meaningful_html_text("\n  Alien   "), true);
+/// assert_eq!(is_meaningful_html_text("  Alien   \n"), true);
+/// assert_eq!(is_meaningful_html_text(""), true);
+/// ```
+pub fn is_meaningful_html_text(text: &str) -> bool {
+    let mut has_newline = false;
+    for c in text.chars() {
+        // If there is a non-whitespace character
+        if !HTML_WHITESPACE_CHARS.contains(&c) {
+            return true;
+        } else if c == '\n' {
+            has_newline = true;
+        }
```

Comment:
Is it useful to add a comment explaining the newline detection?

```suggestion
        } else if c == '\n' {
            // A newline is detected. This is tracked separately, 
            // potentially to differentiate text with just line breaks from other whitespace.
            has_newline = true;
        }
```

Reply:
This is explained in the doc comment for the function.